openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/LiS/man/man2/putpmsg.2s.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) putpmsg.2s.man,v 1.1.4.3 2006/04/05 09:42:18 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:42:18 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" putpmsg.2s.man,v
.\" Revision 1.1.4.3  2006/04/05 09:42:18  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH PUTPMSG 2 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ System Calls"
.SH NAME
.B putpmsg
\- put a band message to a STREAMS character device
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.HP 8
.BI "int " retval " = putpmsg(int " fd ", const struct strbuf *" ctlptr ", const struct strbuf *" dataptr ", int " band ", int " flags );
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.I fd
the file descriptor to which to put the message.
.TP
.I ctlptr
a pointer to a
.B strbuf
structure describing the control part of the message.
.TP
.I dataptr
a pointer to a
.B strbuf
structure describing the data part of the message.
.TP
.I band
the band for the message.
.TP
.I flags
flags for the message.
.\"
.\"
.SH DESCRIPTION
.PP
.B putpmsg()
generates a STREAMS message from the buffer supplied in the specified
.IR ctlptr " and " datapr
for the queue band
.I band
and with options specified by
.I flags
and delivers the message to the stream head associated with the STREAMS
character special file
.IR fd .
.PP
The resulting STREAMS message can contain a control part as specified by
.IR ctlptr ,
a data part as specified by
.IR datptr ,
or both.
The control part, when present, will generate
.BR M_PROTO " or " M_PCPROTO
message blocks.
The data part, when present, will generate
.B M_DATTA
message blocks.
When both control and data parts are present, an
.BR M_PROTO " or " M_PCPROTO
message block followed by one or more
.B M_DATA
message blocks will be generated.
.PP
.I fd
is an open file descriptor for the STREAMS character special file to which the
STREAMS message is to be written.
.PP
.IR ctlptr " and " datptr
point to a
.B strbuf
structure, which contains the following members:
.sp
.nf
\fC\s-2\
struct strbuf {
    int maxlen;         /* Maximum buffer length.  */
    int len;            /* Length of data.  */
    char *buf;          /* Pointer to buffer.  */
};
\s+2\fR
.fi
.TP 12
.I maxlen
Not used by
.BR putpmsg() .
.TP 12
.I len
The length of the information pointed to by
.IR buf .
.TP 12
.I buf
A pointer to the buffer that contains
.I len
bytes of information.
.PP
.I ctlptr
points to a
.B strbuf
structure describing the control part to be included in the message, or
.B NULL
if there is no control part.
The
.I buf
member
of the
.B strbuf
structure pointed to by
.I ctlptr
indiates the start of the information to sent in the control part, and
the
.I len
member indicates the length of the information for the control part.
.PP
.I datptr
points to a
.B strbuf
structure decribing the data part to be included in the message, or
.B NULL
is there is no data part.
The
.I buf
member
of the
.B strbuf
structure pointed to by
.I ctlptr
indiates the start of the information to sent in the data part, and
the
.I len
member indicates the length of the information for the data part.
.I len
can be zero (0) to send zero-length data.
.PP
.I band
specifies the queue band to which to write data when
.I flags
is set to
.BR MSG_BAND .
When flags is set to other than
.BR MSG_BAND ,
.I band
must be set to zero (0).
.PP
.I flags
specifies whether to send a high priority message or a band message.
.I flags
can be one of the following values:
.TP 12
.B MSG_HIPRI
Specifies that a high priority message is to be sent if the request contains a
control part
.IR "" ( ctlptr " is not"
.BR "" NULL ).
This results in an
.B M_PCPROTO
message being placed on the stream head.
.TP 12
.B MSG_BAND
Specifies that a normal priority queue band message is to be sent for the
specified
.IR band .
This results in an
.B M_PROTO
message being placed on the stream head when there is a control part.
When there is only a data part, an
.B M_DATA
message for the specified
.I band
is placed on the stream head.
.\"
.\"
.SH "RETURN VALUE"
.PP
Upon success,
.B putpmsg()
returns zero (0).
.PP
Upon failure,
.B putpmsg()
returns 1 and sets
.I errno
to an appropritate error number.
.\"
.\"
.SH ERRORS
.PP
When
.B putpmsg()
fails, it returns -1 and sets
.I errno
to one of the following:
.TP 12
.\".RB [ EACCES ]
.\".I fd
.\"is a dynamic device, and write permission on the device is denied.
.TP 12
.RB [ EAGAIN ]
The stream head is set for non-blocking operation and
.B putpmsg()
would block: that is, the message is being sent with
.I flags
set to
.B MSG_BAND
and the corresponding
.I band
is flow controlled.
.TP 12
.RB [ EBADF ]
.I fd
is not a valid file descriptor or is not open for reading.
.TP 12
.RB [ EFAULT ]
.IR ctlptr ", " datptr ", " ctlptr->buf " or " datptr->buf
point outside the caller's address space.
.TP 12
.RB [ EINTR ]
A signal arrived before the operation could complete.
.TP 12
.RB [ EINVAL ]
The values of
.IR band " or " flags ,
or the combination of
.IR ctlptr ", " datptr ", and the values of " band " and " flags
were invalid,
such as when
.I flags
is specified as
.B MSG_HIPRI
and
.I band
is non-zero,
or,
.I fd
refers to a stream that is linked under a STREAMS multiplexing driver.
.TP 12
.RB [ EIO ]
I/O error, or,
.I fd
refers to a stream that is open but is in the process of closing, or refers to
a pipe and the other end of the pipe is closed.
.TP 12
.RB [ EISDIR ]
.I fd
refers to a directory.
.TP 12
.RB [ ENODEV ]
.I fd
refers to a device that does not support the
.B putpmsg()
system call.
.TP 12
.RB [ ENOSR ]
STREAMS resources were insufficient to complete the operation.
.TP 12
.RB [ ENOSTR ]
.I fd
refers to a device that has no stream associated with it.
.TP 12
.RB [ ENXIO ]
A
.B M_HANGUP
message arrived at the stream head for the specified
.IR fd .
.TP 12
.RB [ ERANGE ]
The
.I len
of the data part of the message is outside the range
.I q_minpsz
to
.I q_maxpsz
inclusive for the topmost STREAMS module or driver under the stream head, or,
the
.I len
of the control or data part is larger than the maximum sized control or data
message for the STREAMS subsystem.
.PP
Other errors may be returned by
.BR putpmsg() .
If the protocol module sends an
.B M_ERROR
message to the stream head, the error returned on all subsequent write
operations, including
.BR putpsmg() ,
is specified in the
.B M_ERROR
message by the protocol module.  Which errors are returned under what
conditions form part of the necessary documentation of the protocol module.
.\"
.\"
.SH CAVEATS
.PP
.B LiS
.[
LiS
.]
could handle concurrent writes better.  Linux socket code blocks concurrent
writers independently while partial writers are installed agains a socket.
When partial writes are performed for messages using
.BR putpmsg() ,
.B LiS
.[
LiS
.]
should install the partial writer and block all writers from completing the
partial write on a given band until the partial writer completes the write or
the partial writer closes the stream.  This would at least keep multiple
threads of execution from interfering with partial writes.
.\"
.\"
.SH BUGS
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EINVAL ]
instead of
.RB [ EIO ]
when the internal head queue pointer points to a structure that is not a queue
structure, or is mangled.
.PP
.B LiS
.[
LiS
.]
does not return
.RB [ EIO ]
when
.B putpmsg()
is called on a closing stream.
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EAGAIN ]
instead of
.RB [ EINTR ]
when an
.B M_SIG
message causes a signal that interrupts the call to
.B putpmsg()
before any data is written.
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ ENODEV ]
instead of
.RB [ ENOSTR ]
when there is no stream associated with
.IR fd .
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EINVAL ]
instead of
.RB [ ENODEV ]
when
.I fd
refers to a device that does not support the
.B putpmsg()
system call.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR intro (2),
.BR poll (2),
.BR putmsg (2),
.BR write (2),
.BR streamio (2).
.\"
.\"
.SH "CONFORMING TO"
.PP
.BR SVID ", " SUSv2 ", " SUSv3 ", " POSIX .
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for
.B LiS\c
.
.[
LiS
.]
.PP
.B putpmsg()
first appeared in SVR 3.
.[
svr3
.]
.\"
.\"
.XX
.[
magic
.]
.[
svr42
.]
.[
svr4
.]
.[
svr3
.]
.[
svid
.]
.[
xns
.]
.[
xbd5
.]
.[
$LIST$
.]
.TI


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/lib/libc/putpmsg.c
/*****************************************************************************

 @(#) putpmsg.c,v OpenSS7-0_9_2_D_rc2(0.9.2.14) 2005/12/14 16:30:05

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2005  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2005/12/14 16:30:05 by brian

 *****************************************************************************/

#ident "@(#) putpmsg.c,v OpenSS7-0_9_2_D_rc2(0.9.2.14) 2005/12/14 16:30:05"

static char const ident[] =
    "putpmsg.c,v OpenSS7-0_9_2_D_rc2(0.9.2.14) 2005/12/14 16:30:05";

#define _XOPEN_SOURCE 600
#define _REENTRANT
#define _THREAD_SAFE

#include <sys/types.h>
#include <unistd.h>
#include <stropts.h>

#include <pthread.h>
#include <errno.h>

extern void __pthread_testcancel(void);

#pragma weak __pthread_testcancel
#pragma weak pthread_testcancel

void
pthread_testcancel(void)
{
	if (__pthread_testcancel)
		__pthread_testcancel();
	return;
}

/**
 * @fn int putpmsg(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int band, int flags)
 * @ingroup libLiS
 * @brief put a message to a stream band.
 * @param fd a file descriptor representing the stream.
 * @param ctlptr a pointer to a strbuf structure describing the control part of the message.
 * @param datptr a pointer to a strbuf structure describing the data part of the message.
 * @param band the band to which to put the message.
 * @param flags the priority of the message.
 *
 * putpmsg() must contain a thread cancellation point (SUS/XOPEN/POSIX).  In
 * the Linux Threads approach, this function will return EINTR if interrupted
 * by a signal.  When the function returns EINTR, the Linux Threads user
 * should check for cancellation with pthread_testcancel().  Because this
 * function consists of a single system call, asynchronous thread cancellation
 * protection is not required.
 */
static int
__putpmsg(int fd, const struct strbuf *ctlptr, const struct strbuf *datptr, int band, int flags)
{
	int err;
	struct strpmsg args;

	args.ctlbuf = ctlptr ? *ctlptr : ((struct strbuf) {
					  -1, -1, NULL});
	args.databuf = datptr ? *datptr : ((struct strbuf) {
					   -1, -1, NULL});
	args.band = band;
	args.flags = flags;
#if defined HAVE_KMEMB_STRUCT_FILE_OPERATIONS_UNLOCKED_IOCTL
	if ((err = ioctl(fd, I_PUTPMSG, &args)) == -1)
#else
	if ((err = write(fd, &args, LFS_GETMSG_PUTMSG_ULEN)) == -1)
#endif
	{
		int __olderrno;
		/* If we get an EINVAL error back it is likely due to a bad ioctl, in which case
		   this is not a Stream, so we need to check if it is a Stream and fix up the error
		   code.  We get EINTR for a controlling terminal. */
		if ((__olderrno = errno) == EINVAL || __olderrno == EINTR || __olderrno == ENOTTY)
			errno = (ioctl(fd, I_ISASTREAM) == -1) ? ENOSTR : __olderrno;
	}
	return (err);
}

int
putpmsg(int fd, const struct strbuf *ctlptr, const struct strbuf *datptr, int band, int flags)
{
	int ret;

	pthread_testcancel();
	ret = __putpmsg(fd, ctlptr, datptr, band, flags);
	if (ret == -1)
		pthread_testcancel();
	return (ret);
}
